wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/urlfactory/DefaultFactoryUrl.java

```diff
@@ -61,6 +67,39 @@ public String getBranch() {
     return null;
   }
 
+  @Override
+  public @Nullable Optional<String> getCredentials() {
+    if (isNullOrEmpty(url)) {
+      return Optional.empty();
+    }
+    Matcher matcher =
+        compile("https?://((?<username>[^:|@]+)?:?(?<password>[^@]+)?@)?.*").matcher(url);
+    String password = null;
+    String username = null;
+    try {
+      username = matcher.matches() ? matcher.group("username") : null;
+    } catch (IllegalArgumentException e) {
+      // no such group
+    }
+    try {
+      password = matcher.matches() ? matcher.group("password") : null;
+    } catch (IllegalArgumentException e) {
+      // no such group
+    }
+    if (!isNullOrEmpty(username) || !isNullOrEmpty(password)) {
+      return Optional.of(
+          format(
+              "%s:%s",
+              isNullOrEmpty(username) ? "" : username, isNullOrEmpty(password) ? "" : password));
+    }
+    return Optional.empty();
+  }
+
+  public <T extends DefaultFactoryUrl> T withUrl(String url) {
+    this.url = url;
+    return (T) this;
+  }
```

Comment:
I do not think we need @Nullable with Optional, also please consider simplified version of the code - https://github.com/eclipse-che/che-server/pull/451#discussion_r1114270441

using Matcher less readable, more fragile and error prone

Reply:
done